File: Data_Structure/Java/BinarySearchTree.java
import java.util.Scanner;

class BSTNode {
    int data;
    BSTNode left, right;
    BSTNode(int value) {
        data = value;
        left = right = null;
    }
}

class BinarySearchTree {
    BSTNode root;

    BSTNode insert(BSTNode root, int data) {
        if (root == null) return new BSTNode(data);
        if (data < root.data) root.left = insert(root.left, data);
        else if (data > root.data) root.right = insert(root.right, data);
        return root;
    }

    BSTNode search(BSTNode root, int key) {
        if (root == null || root.data == key) return root;
        if (key < root.data) return search(root.left, key);
        else return search(root.right, key);
    }

    void inorder(BSTNode root) {
        if (root != null) {
            inorder(root.left);
            System.out.print(root.data + " ");
            inorder(root.right);
        }
    }

    public static void main(String[] args) {
        BinarySearchTree tree = new BinarySearchTree();
        Scanner sc = new Scanner(System.in);
        int value, key;

        System.out.println("Enter values to insert into BST (-1 to stop):");
        while (true) {
            value = sc.nextInt();
            if (value == -1) break;
            tree.root = tree.insert(tree.root, value);
        }

        System.out.print("Inorder Traversal: ");
        tree.inorder(tree.root);
        System.out.println();

        System.out.print("Enter value to search: ");
        key = sc.nextInt();

        if (tree.search(tree.root, key) != null)
            System.out.println(key + " found in BST.");
        else
            System.out.println(key + " not found in BST.");

        sc.close();
    }
}

File: Data_Structure/Python/BinarySearchTree.py
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

class BST:
    def insert(self, root, data):
        if root is None:
            return Node(data)
        if data < root.data:
            root.left = self.insert(root.left, data)
        elif data > root.data:
            root.right = self.insert(root.right, data)
        return root

    def search(self, root, key):
        if root is None or root.data == key:
            return root
        if key < root.data:
            return self.search(root.left, key)
        else:
            return self.search(root.right, key)

    def inorder(self, root):
        if root:
            self.inorder(root.left)
            print(root.data, end=" ")
            self.inorder(root.right)

# Usage
tree = BST()
root = None

print("Enter values to insert into BST (-1 to stop):")
while True:
    value = int(input())
    if value == -1:
        break
    root = tree.insert(root, value)

print("Inorder Traversal:", end=" ")
tree.inorder(root)
print()

key = int(input("Enter value to search: "))
if tree.search(root, key):
    print(f"{key} found in BST.")
else:
    print(f"{key} not found in BST.")